File: src/main/java/no/fintlabs/integration/IntegrationRepository.java
package no.fintlabs.integration;

import no.fintlabs.integration.model.entities.Integration;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;

import java.util.List;
import java.util.Optional;
import java.util.Set;

public interface IntegrationRepository extends JpaRepository<Integration, Long> {

    boolean existsIntegrationBySourceApplicationIdAndSourceApplicationIntegrationId(
            Long sourceApplicationId,
            String sourceApplicationIntegrationId
    );

    Optional<Integration> findIntegrationBySourceApplicationIdAndSourceApplicationIntegrationId(
            Long sourceApplicationId,
            String sourceApplicationIntegrationId
    );

    List<Integration> findIntegrationsBySourceApplicationIdIn(Set<Long> sourceApplicationIds);

    Page<Integration> findIntegrationsBySourceApplicationIdIn(Set<Long> sourceApplicationIds, Pageable pageable);

}

File: src/main/kotlin/no/novari/flyt/integration/persistence/IntegrationRepository.kt
package no.novari.flyt.integration.persistence

import no.novari.flyt.integration.persistence.entity.Integration
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query

interface IntegrationRepository : JpaRepository<Integration, Long> {
    @Query(
        """
        SELECT DISTINCT integration.sourceApplicationId
        FROM Integration integration
        WHERE integration.sourceApplicationId IS NOT NULL
        """,
    )
    fun findDistinctSourceApplicationIds(): Set<Long>

    fun existsIntegrationBySourceApplicationIdAndSourceApplicationIntegrationId(
        sourceApplicationId: Long,
        sourceApplicationIntegrationId: String,
    ): Boolean

    fun findIntegrationBySourceApplicationIdAndSourceApplicationIntegrationId(
        sourceApplicationId: Long,
        sourceApplicationIntegrationId: String,
    ): Integration?

    fun findIntegrationsBySourceApplicationIdIn(sourceApplicationIds: Set<Long>): List<Integration>

    fun findIntegrationsBySourceApplicationIdIn(
        sourceApplicationIds: Set<Long>,
        pageable: Pageable,
    ): Page<Integration>
}
